keps/sig-node/0029-2018-09-19-node-fencing.md

```diff
@@ -0,0 +1,223 @@
+RWO Volumes. Recovery time for affected Pods and Volumes should be bounded and
+short, allowing workloads to be recovered elsewhere and scale up/down events
+to proceed.
+
+## Motivation
+
+Neither software, the hardware it runs on, nor the networks that connect them
+are perfect.  
+
+Attempts to automatically preserve availabilty in the presence of errors is
+hampered by failures of peer hardware being generally indistinguishable from
+the failure of network and/or software stack.
+
+Using other communication paths can allow the system to know that the peer is
+alive in some capacity, but unless the peer can take steps to recover itself
+in a timely manner, this information is of little use.
+
+Unlike ReplicaSets which can assume the best and spin up new instances, this
+has problematic consequences for anything seeking to provide at-most-one
+semantics.  Especially Stateful Sets which cannot scale up or down while one
+of its members is in a failed state.
+
+Fencing is a well established concept that allows a system to know for sure
+that it is safe to recover these kinds of workloads to other peers.
+Fundamentally it answers the question: _Is this peer capabale of corrupting my
+data?_ With a resounding _No_, and based on that the system can then safely
+initiate recovery.
+
+The traditional non-cloud mechansim for this is to power off the physical
+hardware.  More optimistic implementions power cycle the hardware in the hope
+that it will return to a functional state.  Other more complex systems will
+seek to isolate the peer from the network (so that no additional requests can
+be consumed) and the disk (preventing the double writers problem) in order to
+allow triage activities to take place.
+
+### Experience reports
+
+* Cheng Xing [AttachDetachController fails to force detach on pod deletion](https://github.com/kubernetes/kubernetes/issues/65392) June 2018
+* Clayton Colman [Pod Safety](https://github.com/kubernetes/community/blob/16f88595883a7461010b6708fb0e0bf1b046cf33/contributors/design-proposals/pod-safety.md) October 2016
+
+
+### Goals
+
+* Specify where implementations should look in order to determine if a node requires fencing
+* Specify how implementations should indicate a node needs to be fenced
+* Specify how implementations should indicate a node has been sucessfully fenced
+* Specify how nodes should return to a "normal" state
+* Specify the changes required to Kubernetes that enable automated recovery in the presense of worker failures
+* Provide a sample implementation based on the [Machine](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/machine-api-proposal.md) API 
+
+### Non-Goals
+
+It is explicitly _not_ the goal of this work to:
+
+* Have fencing become a requirement for all Kubernetes deployments
+* Have the community converge on a particular fencing implementation
+
+## Proposal
+
+Three additional NodeConditions and taints:
```

Comment:
You need to talk about the flavor of taint - what is the effect?

It might be good to introduce a real-world example and use that to motivate the design points.

Reply:
Will do